demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c

```diff
@@ -0,0 +1,613 @@
+            {
+                configPRINTF( ( "Failed to retrieve EC Point.\r\n" ) );
+            }
+
+            /* Perform a sign with the generated private key. */
+            xMechanism.mechanism = CKM_ECDSA;
+            xMechanism.pParameter = NULL;
+            xMechanism.ulParameterLen = 0;
+            xResult = pxGlobalFunctionList->C_SignInit( xGlobalSession, &xMechanism, xPrivateKeyHandle );
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Failed to SignInit ECDSA.\r\n" ) );
+            }
+
+            xSignatureLength = sizeof( xSignature );
+            xResult = pxGlobalFunctionList->C_Sign( xGlobalSession, xHashedMessage, SHA256_DIGEST_SIZE, xSignature, &xSignatureLength );
+
+            if( xResult != CKR_OK )
+            {
+                configPRINTF( ( "Failed to ECDSA Sign.\r\n" ) );
+            }
+
+            /* Verify the signature with mbedTLS */
+            mbedtls_ecdsa_init( &xEcdsaContext );
+            mbedtls_ecp_group_init( &xEcdsaContext.grp );
+
+
+            lMbedTLSResult = mbedtls_ecp_group_load( &xEcdsaContext.grp, MBEDTLS_ECP_DP_SECP256R1 );
+
+            if( lMbedTLSResult != 0 )
+            {
+                configPRINTF( ( "mbedTLS failed in setup for signature verification.\r\n" ) );
+            }
+
+            /* The first 2 bytes are for ASN1 type/length encoding. */
+            lMbedTLSResult = mbedtls_ecp_point_read_binary( &xEcdsaContext.grp, &xEcdsaContext.Q, &xEcPoint[ 2 ], xTemplate.ulValueLen - 2 );
+
+            if( lMbedTLSResult != 0 )
+            {
+                configPRINTF( ( "mbedTLS failed in setup for signature verification.\r\n" ) );
+            }
+
+            /* Generate a PK Context for CSR */
+            /* Getting header info from ECKEY type */
+            mbedtls_pk_type_t type_key = MBEDTLS_PK_ECKEY;
+            const mbedtls_pk_info_t * header = mbedtls_pk_info_from_type( type_key );
+
+            /* Creating copy of header to pass in custom sign function (original header is immutable) */
+            mbedtls_pk_info_t * header_copy = pvPortMalloc( sizeof( mbedtls_pk_info_t ) );
+            memcpy( header_copy, header, sizeof( struct mbedtls_pk_info_t ) );
+
+            header_copy->sign_func = &prvPrivateKeySigningCallback;
+
+            /* Initializing PK Context */
+            mbedtls_pk_context pk_cont;
+            mbedtls_pk_init( &pk_cont );
+            ret = mbedtls_pk_setup( &pk_cont, header_copy );
+
+            if( ret != 0 )
+            {
+                configPRINTF( ( "Failed to initialize PK context with given information.\r\n" ) );
+            }
+
+            pk_cont.pk_ctx = &xEcdsaContext;
+
+
+            /* Generating CSR */
+            /* Initializing CSR Context */
+            mbedtls_x509write_csr my_csr;
+            mbedtls_x509write_csr_init( &my_csr );
+            ret = mbedtls_x509write_csr_set_subject_name( &my_csr, "CN=ThingName" );
```

Comment:
I'd suggest highlighting that this thing name is configurable (maybe we pull it from clientcredential, or from a define)

Also, from the template, is the thing name being pulled from the common name or the certificate ID?  For this example, I think common name is the way to go like you have done here :)

Reply:
Added a comment regarding thing name configuration.

Yes, I originally had the thing name pulled from the cert ID in the template -- I just modified it to pull from the common name instead :)